core/primitives/src/chain.rs

```diff
@@ -44,20 +44,37 @@ pub struct SignedShardBlock {
 pub struct ReceiptBlock {
     pub header: SignedShardBlockHeader,
     pub path: MerklePath,
+    // receipts should not be empty
     pub receipts: Vec<ReceiptTransaction>,
+    // hash is the hash of receipts. It is
+    // sufficient to uniquely identify the 
+    // receipt block because of the uniqueness
+    // of nonce in receipts
+    pub hash: CryptoHash,
```

Comment:
This has some issues with serialization:
* `hash` should be excluded from serialization, see: https://serde.rs/attr-skip-serializing.html
* If `ReceiptBlock` was constructed through deserialization `hash` is not going to be computed and will have an incorrect default value. Instead we should compute hash lazy. Meaning `hash` should be `Option<RefCell<CryptoHash>>`, then `Hash::hash` should be checking if it is `None`, and if it is then it computes it. It could have been easier if serde was supporting running custom code upon deserialization, but it currently does not: https://github.com/serde-rs/serde/issues/642

Reply:
Not sure if we want to do this right now. If so, we need to fix `SignedTransaction`, `SignedBeaconBlock`, `SignedBeaconBlockHeader`, `SignedShardBlock`, and `SignedShardBlockHeader` as well.

Reply:
In general, we need to move to computing hashes once, so at some point, we will have to do this modification to the structs that you have listed. However, since we have decided to do it to one struct at the moment, `ReceiptBlock`, then why not do it correctly now. Especially since, the modification that I proposed does not add much code. Had it involved adding a lot of code, I would have agreed with your argument.

Reply:
Since the `SignedBlock` trait requires `Sync`, we cannot use `RefCell` in `ReceiptBlock`. I tried to use `Mutex` but since `Mutex` does not implement `Clone`, it also cannot satisfy the requirement of `SignedBlock`.

Reply:
`SignedBlock` does not need neither `Send` nor `Sync`. I removed both of them and the project still compiles.

Reply:
It would fail here https://github.com/nearprotocol/nearcore/blob/3af2a57ba1029896013b4e7f523a530fb4cb1430/node/coroutines/src/client_task.rs#L412 with `RefCell`.

Reply:
I see, it would have to be `Arc<RefCell<Option<CryptoHash>>`. I think it is complex enough to be postponed to a different PR.